src/main/scala/com/gu/sfl/lib/SavedArticlesMerger.scala

```diff
@@ -0,0 +1,68 @@
+package com.gu.sfl.lib
+
+import com.gu.sfl.Logging
+import com.gu.sfl.exception.{MaxSavedArticleTransgressionError, SavedArticleMergeError}
+import com.gu.sfl.model._
+import com.gu.sfl.persisitence.SavedArticlesPersistence
+
+import scala.annotation.tailrec
+import scala.util.{Failure, Success, Try}
+
+case class SavedArticlesMergerConfig(maxSavedArticlesLimit: Int)
+
+trait SavedArticlesMerger {
+  def updateWithRetryAndMerge(userId: String, savedArticles: SavedArticles): Try[Option[SyncedPrefs]]
+}
+
+class SavedArticlesMergerImpl(savedArticlesMergerConfig: SavedArticlesMergerConfig, savedArticlesPersistence: SavedArticlesPersistence) extends SavedArticlesMerger with Logging {
+
+  val maxSavedArticlesLimit: Int = savedArticlesMergerConfig.maxSavedArticlesLimit
+
+  private def persistMergedArticles(userId: String, articles: SavedArticles)( persistOperation: (String, SavedArticles) => Try[Option[SavedArticles]] ): Try[Option[SyncedPrefs]] = persistOperation(userId, articles) match {
+    case Success(Some(articles)) =>
+      logger.debug(s"success persisting articles for ${userId}")
+      Success(Some(SyncedPrefs(userId, Some(articles))))
+
+    case Failure(e) =>
+      logger.debug(s"Error persisting articles for ${userId}. Error: ${e.getMessage}")
+      Failure(SavedArticleMergeError("Could not update articles"))
+  }
+
+  override def updateWithRetryAndMerge(userId: String, savedArticles: SavedArticles): Try[Option[SyncedPrefs]] = {
+
+    @tailrec
+    def loop(articles: SavedArticles, retries: Int): Try[Option[SyncedPrefs]] = {
+
+      if(retries == 0) {
+        logger.debug(s"Failed to merge saved articles for user: $userId")
+        Failure(SavedArticleMergeError("Conflicting version in savedArticles") )
+      } else {
+        savedArticlesPersistence.read(userId) match {
+          case Success(Some(currentArticles)) =>
+            val mergedArticles = Merge.mergeListBy(currentArticles.articles, articles.articles)(_.id)
+            if(currentArticles.version == articles.version) {
```

Comment:
Is this a candidate for a conditional expression? https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Expressions.ConditionExpressions.html

Reply:
Almost, but not really. We're dealing with a scenario where a device is saving to a record which has been updated by another device ( ie, the version in the current request is different from the currently persisted version ) where the version is the same we know this isn't case. Where it differs, we take the currently saved record and attemp to merge the article list on the request with the list currently persisted. We then try to to save the merged list with the currently (as in when we started merging ) persisted version value. If we do a conditional update, we can't really do this